opm/autodiff/FullyImplicitBlackoilSolver_impl.hpp

```diff
@@ -630,10 +630,27 @@ namespace detail {
         //    Note that some of the complexity of this part is due to the function
         //    taking std::vector<double> arguments, and not Eigen objects.
         const int nperf = wells().well_connpos[wells().number_of_wells];
+        const int nw = wells().number_of_wells;
+        const int np = wells().number_of_phases;
         const std::vector<int> well_cells(wells().well_cells, wells().well_cells + nperf);
+
+        // Compute the average pressure in each well block
+        const V perf_press = Eigen::Map<const V>(xw.perfPress().data(), nperf);
+        V avg_press = perf_press*0;
+        for (int w = 0; w < nw; ++w) {
+            for (int perf = wells().well_connpos[w]; perf < wells().well_connpos[w+1]; ++perf) {
+                const double p_above = perf == wells().well_connpos[w] ? state.bhp.value()[w] : perf_press[perf - 1];
```

Comment:
This (and I guess some already existing code) seems to require that the bhp reference depth is at most that of the first perforation to work well (although it should not explode otherwise), do we have any tests where the bhp reference depths are somewhere further down along the well bore?